posts/2024-04-29-check-cfg.md

```diff
@@ -0,0 +1,115 @@
+
+Some crates use custom cfgs that they either expected from the environment (`RUSTFLAGS`or other means) or is enabled by some logic in the crate `build.rs`. For those crates Cargo provides a new instruction: [`cargo::rustc-check-cfg`](TODO)[^2] (or `cargo:rustc-check-cfg` for older Cargo version).
+
+[^2]: `cargo::rustc-check-cfg` take action since Rust 1.80, between Rust 1.77 and Rust 1.79 *(included)* it silently did nothing and before that (so Rust 1.76 and below) it warned when used without the unstable Cargo `-Zcheck-cfg`.
+
+The syntax to use is described in the [rustc book](https://doc.rust-lang.org/rustc/index.html) section [checking configuration](TODO), but in a nut shell the basic syntax of `--check-cfg` is:
+
+> `cfg(name, values("value1", "value2", ..., "valueN"))`
+
+Note that every custom cfgs must always be expected, regardless if the cfg is active or not!
+
+### `build.rs` example
+
+`build.rs`:
+```rust
+fn main() {
+    println!("cargo::rustc-check-cfg=cfg(has_foo)"); // <-- new with Cargo 1.80
+    if has_foo() {
+        println!("cargo::rustc-cfg=has_foo");
+    }
+}
+```
+
+> Note: Each `cargo::rustc-cfg` must always have a accompanying _unconditional_ `cargo::rustc-check-cfg` directive otherwise you risk having warnings like this:`unexpected cfg condition name: has_foo`.
+
+### Equivalence table
+
+| `cargo::rustc-cfg`      | `cargo::rustc-check-cfg`                       |
+|-------------------------|------------------------------------------------|
+| `foo`                   | `cfg(foo)` or `cfg(foo, values(none()))`       |
+| `foo=""`                | `cfg(foo, values(""))`                         |
+| `foo="bar"`             | `cfg(foo, values("bar"))`                      |
+| `foo="1"` and `foo="2"` | `cfg(foo, values("1", "2"))`                   |
+| `foo="1"` and `bar="2"` | `cfg(foo, values("1"))` and `cfg(bar, values("2"))` |
+| `foo` and `foo="bar"`   | `cfg(foo, values(none(), "bar"))`              |
+
+More details can be found on the [`rustc` book](TODO).
+
+## Frequently asked questions
+
+### Can it be disabled?
+
+No, it's **always on** for Cargo and _cannot_ be disabled.
+
+### How does it interact with the `RUSTFLAGS` env ?
+
+You should be able to use the `RUSTFLAGS` environment variable like it was before.
+*Currently `--cfg` arguments are not checked, only usage in code are.*
+
+This means that doing `RUSTFLAGS="--cfg tokio_unstable" cargo check` will not report any warnings, unless `tokio_unstable` is used within your local crates, in which case crate author will need to make sure that that custom cfg is expected with `cargo::rustc-check-cfg` in the `build.rs` of that crate.
+
+### How to expect custom cfgs without a `build.rs` ?
+
+There is not **currently no way** to expect a custom cfg other than with `cargo::rustc-check-cfg` in a `build.rs`.
```

Comment:
```suggestion
There is **currently not any way** to expect a custom cfg other than with `cargo::rustc-check-cfg` in a `build.rs`.
```

Reply:
maybe you need a section or at least a link for how to use this when using rustc without cargo, e.g. in the Linux Kernel.

Reply:
This blog post is very much centered around Cargo users (it is even posted by the Cargo team). So a full section for non-Cargo users feels overkill and out of place.

However we could have question that basicly just links to the rustc docs, which contains all the details for that.

Btw, the Linux Kernel tested the feature before stabilization and is planning on using it.

Reply:
maybe phrase it something like:

> Cargo achieves this by using the `--whatever-its-called` rustc options

which allows you to smoothly link to rustc's docs there.

Reply:
I just pushed a new question/section regarding non-Cargo based build system, which I think should address your remark.

Feel free to leave review comments on it.